File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    static List<String> list;
    public List<String> generateParenthesis(int n) {
        list=new ArrayList<>();
        solve(n,0,0,"");
        return list;
        
    }
    public static void solve(int n, int open , int closed, String ans){
        if(open==n && closed==n){
            list.add(ans);
            return;
        }

        if(open<n){
            solve(n,open+1,closed,ans+"(");
        }

        if(closed<open){
            solve(n,open,closed+1,ans+")");
        }


    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> v;
    vector<string> generateParenthesis(int n) {
        
        generate(n,"",0,0);
        return v;
        
    }
    void generate(int n, string ans, int open, int close){
        if(open==n && close==n){
            v.push_back(ans);
            return;
        }

        if(open<n){
            generate(n,ans+"(" ,open+1,close);
        }
        if(close<open){
          generate(n,ans+")",open,close+1);
        }
    }
};